crates/json-abi/src/abi.rs

```diff
@@ -608,3 +649,82 @@ impl<'de> Visitor<'de> for ContractObjectVisitor {
         Ok(ContractObject { abi, bytecode, deployed_bytecode })
     }
 }
+
+/// A [crate::Visitor] implementation that collects a mapping from struct name to set of components
+/// various structs with such name have.
+#[derive(Debug, Default, Clone)]
+struct Collector<'a> {
+    unique_struct_components: BTreeMap<&'a str, BTreeSet<Vec<&'a str>>>,
```

Comment:
perhaps worth using smallvec here for the `Vec` because we only expect a few projections?